File: src/challenges/python/problems/image_smoother_test.py
# https://leetcode.com/problems/image-smoother

from core.problem_base import *

class ImageSmoother(ProblemBase):
    def Solution(self, img: list[list[int]]) -> list[list[int]]:
        directions = [ 
            [ -1, -1 ], [ -1, 0 ], [ -1, 1 ], 
            [  0, -1 ], [  0, 0 ], [  0, 1 ], 
            [  1, -1 ], [  1, 0 ], [  1, 1 ] 
        ]

        count = len(directions)
        rows = len(img)
        cols = len(img[0])
        result = [0] * rows

        for row in range(rows):
            result[row] = [0] * cols
            for col in range(cols):
                neighbors = 0

                for i in range(count):
                    y = row + directions[i][0]
                    x = col + directions[i][1]

                    if y >= 0 and y < rows and x >= 0 and x < cols:
                        result[row][col] += img[y][x]
                        neighbors += 1

                result[row][col] //= neighbors

        return result

if __name__ == '__main__':
    TestGen(ImageSmoother) \
        .Add(lambda tc: tc.Param([[100,200,100],[200,50,200],[100,200,100]]).Result([[137,141,137],[141,138,141],[137,141,137]])) \
        .Add(lambda tc: tc.Param([[1,1,1],[1,0,1],[1,1,1]]).Result([[0,0,0],[0,0,0],[0,0,0]])) \
        .Run()

File: src/rust/challenges/src/problems/image_smoother_test.rs
// https://leetcode.com/problems/image-smoother

pub fn solution(img: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let directions = vec![
        vec![-1, -1], vec![-1, 0], vec![-1, 1],
        vec![0, -1], vec![0, 0], vec![0, 1],
        vec![1, -1], vec![1, 0], vec![1, 1],
    ];

    let count = directions.len();
    let rows = img.len();
    let cols = img[0].len();
    let mut result = vec![vec![0; cols]; rows];

    for row in 0..rows {
        for col in 0..cols {
            let mut neighbors = 0;

            for i in 0..count {
                let y = row as i32 + directions[i][0];
                let x = col as i32 + directions[i][1];

                if y >= 0 && y < rows as i32 && x >= 0 && x < cols as i32 {
                    result[row][col] += img[y as usize][x as usize];
                    neighbors += 1;
                }
            }

            result[row][col] /= neighbors;
        }
    }

    result
}